File: java/src/mgap.java
import java.util.Scanner;

/**
 * @author ekarpov
 */
public class mgap {
    private static class MaxGapMatrix extends LevMatrix {
        private class MaxGapCell extends Cell {
            private int gaps;

            private MaxGapCell(int x, int y, int weight, MaxGapCell prevCell) {
                super(x, y, weight, null, null, prevCell);
                gaps = prevCell != null ? prevCell.gaps + (prevCell.isLeftTo(this) || prevCell.isOnTopOf(this) ? 1 : 0) : 0;
            }
        }

        private MaxGapMatrix(String s, String t) throws InterruptedException {
            super(s, t);
        }

        protected Cell createCell(int x, int y, int weight, Cell prevCell) {
            return new MaxGapCell(x, y, weight, (MaxGapCell) prevCell);
        }

        protected Cell buildCell(Cell diag, Cell top, Cell left, int x, int y) {
            if (top == null && left == null) return createCell(x, y, 0, null);
            else if (top == null) return createCell(x, y, left.weight + getWeight(GAP_SYMBOL, ' '), left);
            else if (left == null) return createCell(x, y, top.weight + getWeight(GAP_SYMBOL, ' '), top);
            else {
                final int topWeight = top.weight + getWeight(GAP_SYMBOL, ' ');
                final int leftWeight = left.weight + getWeight(GAP_SYMBOL, ' ');
                final int diagWeight = diag.weight + getWeight(s.charAt(x), t.charAt(y));

                if (diagWeight > leftWeight && diagWeight > topWeight)
                    return createCell(x, y, diagWeight, diag);
                else if (leftWeight == topWeight) {
                    if (((MaxGapCell) left).gaps > ((MaxGapCell) top).gaps) {
                        return createCell(x, y, leftWeight, left);
                    } else {
                        return createCell(x, y, topWeight, top);
                    }
                } else if (leftWeight > topWeight) {
                    return createCell(x, y, leftWeight, left);
                } else {
                    return createCell(x, y, topWeight, top);
                }
            }
        }
    }

    public static void main(final String[] args) throws InterruptedException {
        final Scanner scanner = new Scanner(System.in);
        final String s = scanner.nextLine();
        final String t = scanner.nextLine();
        final MaxGapMatrix matrix = new MaxGapMatrix(s, t);
        System.out.println(((MaxGapMatrix.MaxGapCell) matrix.getLastCell()).gaps);
    }
}

File: python/mgap.py
from python.lev_matrix import *
from python.timer import Timer


class MaxGapMatrixCell(LevMatrixCell):
    def __init__(self, x, y, weight, s, t, prevCell):
        super().__init__(x, y, weight, s, t, prevCell, True) # suppress prefixes calculation
        if prevCell:
            self.gaps = prevCell.gaps + (1 if prevCell.isOnTopOf(self) or prevCell.isLeftTo(self) else 0)
        else:
            self.gaps = 0


class MaxGapLevMatrix(LevMatrix):
    def __init__(self, s, t):
        super().__init__(s, t, cellFactory=self.createCell,
            weight=lambda x, y: -1 if x == GAP_SYMBOL or y == GAP_SYMBOL else 1 if x == y else -10000, cleanup=True)

    def buildCell(self, diag, top, left, x, y):
        if top is None and left is None:
            return self.cellFactory(x, y, 0)
        elif top is None:
            return self.cellFactory(x, y, left.weight + self.weight(GAP_SYMBOL, None), left)
        elif left is None:
            return self.cellFactory(x, y, top.weight + self.weight(GAP_SYMBOL, None), top)
        else:
            topWeight = top.weight + self.weight(GAP_SYMBOL, None)
            leftWeight = left.weight + self.weight(GAP_SYMBOL, None)
            diagWeight = diag.weight + self.weight(self.s[x], self.t[y])

            if diagWeight > leftWeight and diagWeight > topWeight:
                return self.cellFactory(x, y, diagWeight, diag)
            elif leftWeight == topWeight:
                if left.gaps > top.gaps:
                    return self.cellFactory(x, y, leftWeight, left)
                else:
                    return self.cellFactory(x, y, topWeight, top)
            elif leftWeight > topWeight:
                return self.cellFactory(x, y, leftWeight, left)
            else:
                return self.cellFactory(x, y, topWeight, top)

    def createCell(self, x, y, weight, prevCell=None):
        return MaxGapMatrixCell(x, y, weight, self.s, self.t, prevCell)


if __name__ == '__main__':
    s = input()
    t = input()
    with Timer() as perf:
        matrix = MaxGapLevMatrix(s, t)
        print(matrix.getLastCell().gaps)
    print(perf)
